pkg/pg/collectors.go

```diff
@@ -817,3 +817,350 @@ func PGStatCheckpointer(pgPool *pgxpool.Pool) func(ctx context.Context, state *a
+// It tracks heap blocks processed, index vacuum operations, and dead tuple counts for tables currently being vacuumed.
+// Note: This collector only returns data for tables that are actively being vacuumed at the time of collection.
+//
+// Phase mapping:
+// 0 = initializing
+// 1 = scanning heap
+// 2 = vacuuming indexes
+// 3 = vacuuming heap
+// 4 = cleaning up indexes
+// 5 = truncating heap
+// 6 = performing final cleanup
+// -1 = unknown phase
+func PGStatProgressVacuum(pgPool *pgxpool.Pool) func(ctx context.Context, state *agent.MetricsState) error {
+	return func(ctx context.Context, state *agent.MetricsState) error {
+		var jsonResult *string
+		err := utils.QueryRowWithPrefix(pgPool, ctx, PGStatProgressVacuumQuery).Scan(&jsonResult)
+		if err != nil {
+			return err
+		}
+
+		// If no vacuums are running, jsonResult will be NULL
+		if jsonResult == nil {
+			// Clear the cache and don't emit any metrics
+			state.Cache.PGStatProgressVacuum = nil
+			return nil
+		}
+
+		var vacuumStats map[string]utils.PGStatProgressVacuum
+		err = json.Unmarshal([]byte(*jsonResult), &vacuumStats)
+		if err != nil {
+			return err
+		}
+
+		// Emit current snapshot values (these represent active vacuums)
+		phaseMap := make(map[string]int64)
+		heapBlksTotalMap := make(map[string]int64)
+		heapBlksScannedMap := make(map[string]int64)
+		heapBlksVacuumedMap := make(map[string]int64)
+		indexVacuumCountMap := make(map[string]int64)
+
+		for tableKey, stats := range vacuumStats {
+			phaseMap[tableKey] = stats.Phase
+			heapBlksTotalMap[tableKey] = stats.HeapBlksTotal
+			heapBlksScannedMap[tableKey] = stats.HeapBlksScanned
+			heapBlksVacuumedMap[tableKey] = stats.HeapBlksVacuumed
+			indexVacuumCountMap[tableKey] = stats.IndexVacuumCount
+		}
+
+		metricsToEmit := []struct {
+			metric metrics.MetricDef
+			value  any
+		}{
+			{metrics.PGVacuumPhase, phaseMap},
+			{metrics.PGVacuumHeapBlksTotal, heapBlksTotalMap},
+			{metrics.PGVacuumHeapBlksScanned, heapBlksScannedMap},
+			{metrics.PGVacuumHeapBlksVacuumed, heapBlksVacuumedMap},
+			{metrics.PGVacuumIndexVacuumCount, indexVacuumCountMap},
+		}
+
+		for _, m := range metricsToEmit {
+			err = EmitMetric(state, m.metric, m.value)
+			if err != nil {
+				return err
+			}
+		}
+
+		state.Cache.PGStatProgressVacuum = vacuumStats
+		return nil
+	}
+}
+
+// PGOldTransactions collects information about old transactions and replication slots that may be holding back vacuum.
+// It tracks:
+// 1. Top 5 oldest queries with backend_xmin age > 100M
+// 2. Top 5 idle in transaction sessions with backend_xmin age > 15M
+// 3. Uncommitted prepared transactions
+// 4. Replication slots ordered by xmin age
+// 5. Inactive replication slots
+func PGOldTransactions(pgPool *pgxpool.Pool) func(ctx context.Context, state *agent.MetricsState) error {
+	return func(ctx context.Context, state *agent.MetricsState) error {
+		// 1. Top 5 oldest queries with backend_xmin age > 100M
+		oldestTransactionQuery := `
```

Comment:
I like having all the queries outside of the functions if possible, which I think it is here